File: src/start-chain/prd-generator.ts
/**
 * PRD Generator for RWM Puppet Master
 * 
 * Transforms ParsedRequirements into PRD structure following STATE_FILES.md Section 3.3.
 * Supports both AI-powered generation (via platform runners) and rule-based fallback.
 * 
 * See REQUIREMENTS.md Section 5.2 for PRD generation requirements.
 * See ROADMAP.md 6.2.2 for AI platform integration requirements.
 */

import type { ParsedRequirements, ParsedSection } from '../types/requirements.js';
import type { PRD, Phase, Task, Subtask, PRDMetadata, ItemStatus, SourceRef } from '../types/prd.js';
import type { Criterion, TestPlan } from '../types/tiers.js';
import type { PuppetMasterConfig } from '../types/config.js';
import type { Platform } from '../types/config.js';
import type { ExecutionRequest } from '../types/platforms.js';
import { PlatformRegistry } from '../platforms/registry.js';
import { QuotaManager } from '../platforms/quota-manager.js';
import { UsageTracker } from '../memory/usage-tracker.js';
import { buildPrdPrompt } from './prompts/prd-prompt.js';
import {
  detectDocumentStructure,
  StructureDetectionError,
  type DocumentStructure,
  type StructureDetectorOptions,
} from './structure-detector.js';
import { CriterionClassifier } from './criterion-classifier.js';
import { createHash } from 'crypto';

/**
 * Options for PRD generation.
 */
export interface PrdGeneratorOptions {
  /** Project name for the PRD */
  projectName: string;
  /** Maximum number of subtasks per task (default: 5) */
  maxSubtasksPerTask?: number;
  /** Maximum number of tasks per phase (default: 10) */
  maxTasksPerPhase?: number;
  /** Options for structure detection (default: use defaults from structure-detector) */
  structureDetectorOptions?: StructureDetectorOptions;
  /** FIX 2: Interview assumptions to include in PRD generation prompt */
  interviewAssumptions?: string[];
}

/**
 * Generator that transforms parsed requirements into PRD structure.
 * Supports AI-powered generation with rule-based fallback.
 */
export class PrdGenerator {
  private readonly projectName: string;
  private readonly maxSubtasksPerTask: number;
  private readonly maxTasksPerPhase: number;
  private readonly platformRegistry?: PlatformRegistry;
  private readonly quotaManager?: QuotaManager;
  private readonly config?: PuppetMasterConfig;
  private readonly usageTracker?: UsageTracker;
  private readonly structureDetectorOptions: StructureDetectorOptions;
  private readonly criterionClassifier: CriterionClassifier;
  /** FIX 2: Interview assumptions to include in PRD generation */
  private readonly interviewAssumptions: string[];

  constructor(
    options: PrdGeneratorOptions,
    platformRegistry?: PlatformRegistry,
    quotaManager?: QuotaManager,
    config?: PuppetMasterConfig,
    usageTracker?: UsageTracker
  ) {
    this.projectName = options.projectName;
    this.maxSubtasksPerTask = options.maxSubtasksPerTask ?? 5;
    this.maxTasksPerPhase = options.maxTasksPerPhase ?? 10;
    this.platformRegistry = platformRegistry;
    this.quotaManager = quotaManager;
    this.config = config;
    this.usageTracker = usageTracker;
    this.structureDetectorOptions = options.structureDetectorOptions ?? {};
    this.criterionClassifier = new CriterionClassifier();
    this.interviewAssumptions = options.interviewAssumptions ?? [];
  }

  /**
   * Main entry point: generates a PRD from parsed requirements using AI.
   * Falls back to rule-based generation if AI is unavailable or quota exhausted.
   * 
   * @param parsed - Parsed requirements document
   * @param useAI - Whether to attempt AI generation (default: true)
   * @returns Generated PRD structure
   */
  async generateWithAI(parsed: ParsedRequirements, useAI: boolean = true): Promise<PRD> {
    // If AI dependencies not provided, use fallback
    if (!useAI || !this.platformRegistry || !this.quotaManager || !this.config || !this.usageTracker) {
      return this.generate(parsed);
    }

    // Determine platform (use step-specific config or fallback to phase tier)
    // Config resolution order (P1-T04):
    // 1. config.startChain.prd.platform/model (step-specific)
    // 2. config.tiers.phase.platform/model (default phase tier)
    const stepConfig = this.config.startChain?.prd;
    const platform = stepConfig?.platform || this.config.tiers.phase.platform;
    const model = stepConfig?.model || this.config.tiers.phase.model;

    try {
      // Check quota before proceeding
      const canProceed = await this.quotaManager.canProceed(platform);
      if (!canProceed.allowed) {
        console.warn(`[PRD Generation] Quota exhausted for ${platform}: ${canProceed.reason}. Using rule-based fallback.`);
        return this.generate(parsed);
      }

      // Build prompt (FIX 2: include interview assumptions)
      const prompt = buildPrdPrompt(parsed, this.projectName, this.interviewAssumptions);

      // Get platform runner
      const runner = this.platformRegistry.get(platform);
      if (!runner) {
        console.warn(`[PRD Generation] Platform runner not available for ${platform}. Using rule-based fallback.`);
        return this.generate(parsed);
      }

      // Execute AI request
      const request: ExecutionRequest = {
        prompt,
        model,
        workingDirectory: this.config.project.workingDirectory,
        nonInteractive: true,
        timeout: 300_000, // 5 minutes for PRD generation
      };

      const startTime = Date.now();
      const result = await runner.execute(request);
      const duration = Date.now() - startTime;

      // Record usage
      await this.usageTracker.track({
        platform,
        action: 'prd_generation',
        tokens: result.tokensUsed || 0,
        durationMs: duration,
        success: result.success,
      });

      if (!result.success) {
        console.warn(`[PRD Generation] AI execution failed (exit code ${result.exitCode}). Using rule-based fallback.`);
        return this.generate(parsed);
      }

      // Parse JSON response
      try {
        const prd = this.parsePrdJson(result.output);
        
        // Validate basic structure
        if (!prd.project || !prd.phases || !Array.isArray(prd.phases)) {
          throw new Error('Invalid PRD structure: missing required fields');
        }

        console.log(`[PRD Generation] Successfully generated PRD using ${platform} (${prd.phases.length} phases)`);
        return prd;
      } catch (parseError) {
        console.warn(`[PRD Generation] Failed to parse AI response: ${parseError instanceof Error ? parseError.message : String(parseError)}`);
        console.warn(`[PRD Generation] AI output preview: ${result.output.substring(0, 200)}...`);
        console.warn(`[PRD Generation] Using rule-based fallback.`);
        return this.generate(parsed);
      }
    } catch (error) {
      console.warn(`[PRD Generation] AI generation error: ${error instanceof Error ? error.message : String(error)}. Using rule-based fallback.`);
      return this.generate(parsed);
    }
  }

  /**
   * Rule-based PRD generation (fallback method).
   * This is the original implementation that uses heuristics to transform
   * parsed requirements into PRD structure.
   *
   * Uses structure detection to correctly handle documents with H1 title + H2 sections.
   *
   * @param parsed - Parsed requirements document
   * @returns Generated PRD structure
   * @throws StructureDetectionError if large document produces insufficient phases
   */
  generate(parsed: ParsedRequirements): PRD {
    const now = new Date().toISOString();

    // Detect document structure to determine correct phase sections
    const structure = detectDocumentStructure(
      parsed.sections,
      parsed.rawText,
      this.structureDetectorOptions
    );

    // Log structure detection for debugging
    console.log(
      `[PRD Generation] Structure detected: ${structure.type}, ` +
      `${structure.metrics.phasesCount} phases, ` +
      `${structure.metrics.headingsCount} headings, ` +
      `${structure.metrics.bulletsCount} bullets`
    );

    // Use detected phase sections instead of raw top-level sections
    const phases = this.generatePhases(structure.phaseSections, parsed);
    const metadata = this.calculateMetadata(phases);

    // Use detected title if available, otherwise fall back to parsed title
    const description = structure.title || parsed.title || 'Generated from requirements';

    return {
      project: this.projectName,
      version: '1.0.0',
      createdAt: now,
      updatedAt: now,
      branchName: 'ralph/main',
      description,
      phases,
      metadata,
    };
  }

  /**
   * Generates phases from top-level sections.
   */
  generatePhases(sections: ParsedSection[], parsed: ParsedRequirements): Phase[] {
    const now = new Date().toISOString();
    return sections.map((section, index) => {
      const phaseId = this.generatePhaseId(index + 1);
      const tasks = this.generateTasks(phaseId, section, index + 1, parsed);
      const acceptanceCriteria = this.extractAcceptanceCriteria(section.content, phaseId);
      const testPlan = this.createTestPlan(section.content);
      const sourceRefs = this.generateSourceRefs(section, parsed);

      return {
        id: phaseId,
        title: section.title,
        description: section.content,
        status: 'pending' as ItemStatus,
        priority: 1,
        acceptanceCriteria,
        testPlan,
        tasks,
        sourceRefs,
        createdAt: now,
        notes: '',
      };
    });
  }

  /**
   * Generates tasks from a section's children.
   */
  generateTasks(phaseId: string, section: ParsedSection, phaseIndex: number, parsed: ParsedRequirements): Task[] {
    const now = new Date().toISOString();
    const taskSections = section.children.slice(0, this.maxTasksPerPhase);

    return taskSections.map((taskSection, taskIndex) => {
      const taskId = this.generateTaskId(phaseIndex, taskIndex + 1);
      const subtasks = this.generateSubtasks(taskId, taskSection.content, taskIndex + 1, phaseIndex, parsed, section);
      const acceptanceCriteria = this.extractAcceptanceCriteria(taskSection.content, taskId);
      const testPlan = this.createTestPlan(taskSection.content);
      const sourceRefs = this.generateSourceRefs(taskSection, parsed, section);

      return {
        id: taskId,
        phaseId,
        title: taskSection.title,
        description: taskSection.content,
        status: 'pending' as ItemStatus,
        priority: 1,
        acceptanceCriteria,
        testPlan,
        subtasks,
        sourceRefs,
        createdAt: now,
        notes: '',
      };
    });
  }

  /**
   * Generates subtasks by breaking task content into smaller chunks.
   */
  generateSubtasks(
    taskId: string,
    content: string,
    taskIndex: number,
    phaseIndex: number,
    parsed: ParsedRequirements,
    parentSection: ParsedSection
  ): Subtask[] {
    const now = new Date().toISOString();
    const lines = content.split('\n').filter(line => line.trim().length > 0);
    
    // Split content into chunks (simple heuristic: maxSubtasksPerTask chunks)
    const chunkSize = Math.max(1, Math.ceil(lines.length / this.maxSubtasksPerTask));
    const chunks: string[] = [];
    
    for (let i = 0; i < lines.length; i += chunkSize) {
      const chunk = lines.slice(i, i + chunkSize).join('\n');
      chunks.push(chunk);
    }

    // Limit to maxSubtasksPerTask
    const limitedChunks = chunks.slice(0, this.maxSubtasksPerTask);

    return limitedChunks.map((chunk, subtaskIndex) => {
      const subtaskId = this.generateSubtaskId(phaseIndex, taskIndex, subtaskIndex + 1);
      const acceptanceCriteria = this.extractAcceptanceCriteria(chunk, subtaskId);
      const testPlan = this.createTestPlan(chunk);
      // For subtasks, use the parent section (task section) as the source reference
      const sourceRefs = this.generateSourceRefs(parentSection, parsed);

      return {
        id: subtaskId,
        taskId,
        title: `Subtask ${subtaskIndex + 1}`,
        description: chunk,
        status: 'pending' as ItemStatus,
        priority: 1,
        acceptanceCriteria,
        testPlan,
        iterations: [],
        maxIterations: 3,
        sourceRefs,
        createdAt: now,
        notes: '',
      };
    });
  }

  /**
   * Generates a phase ID in format PH-XXX.
   */
  generatePhaseId(index: number): string {
    return `PH-${String(index).padStart(3, '0')}`;
  }

  /**
   * Generates a task ID in format TK-XXX-XXX.
   */
  generateTaskId(phaseIndex: number, taskIndex: number): string {
    return `TK-${String(phaseIndex).padStart(3, '0')}-${String(taskIndex).padStart(3, '0')}`;
  }

  /**
   * Generates a subtask ID in format ST-XXX-XXX-XXX.
   */
  generateSubtaskId(phaseIndex: number, taskIndex: number, subtaskIndex: number): string {
    return `ST-${String(phaseIndex).padStart(3, '0')}-${String(taskIndex).padStart(3, '0')}-${String(subtaskIndex).padStart(3, '0')}`;
  }

  /**
   * Extracts acceptance criteria from content.
   * Looks for bullet points and numbered lists.
   */
  extractAcceptanceCriteria(content: string, itemId: string): Criterion[] {
    const criteria: Criterion[] = [];
    const lines = content.split('\n');

    let criterionIndex = 1;
    for (const line of lines) {
      const trimmed = line.trim();
      
      // Check for bullet points (-, *, •) or numbered lists
      const bulletMatch = trimmed.match(/^[-*•]\s+(.+)$/);
      const numberedMatch = trimmed.match(/^\d+[.)]\s+(.+)$/);
      
      if (bulletMatch || numberedMatch) {
        const description = bulletMatch ? bulletMatch[1] : numberedMatch![1];
        const baseType = this.criterionClassifier.classifyAcceptanceCriterion(description);
        const provisional: Criterion = {
          id: `${itemId}-AC-${String(criterionIndex).padStart(3, '0')}`,
          description,
          type: baseType,
          target: '',
        };
        const target = this.criterionClassifier.generateVerificationTarget(provisional);
        const finalType = target.startsWith('AI_VERIFY:') ? 'ai' : baseType;
        criteria.push({
          ...provisional,
          type: finalType,
          target,
        });
        criterionIndex++;
      }
    }

    // If no criteria found, create a generic one
    if (criteria.length === 0) {
      const description = 'Implementation complete';
      const provisional: Criterion = {
        id: `${itemId}-AC-001`,
        description,
        type: 'ai',
        target: '',
      };
      const target = this.criterionClassifier.generateVerificationTarget(provisional);
      criteria.push({
        ...provisional,
        target,
      });
    }

    return criteria;
  }

  /**
   * Creates a default test plan from content.
   * Initially returns empty test plan (can be enhanced later).
   */
  createTestPlan(_content: string): TestPlan {
    return {
      commands: [],
      failFast: true,
    };
  }

  /**
   * Calculates metadata counts from phases.
   */
  calculateMetadata(phases: Phase[]): PRDMetadata {
    let totalTasks = 0;
    let totalSubtasks = 0;
    let completedPhases = 0;
    let completedTasks = 0;
    let completedSubtasks = 0;

    for (const phase of phases) {
      if (phase.status === 'passed') {
        completedPhases++;
      }

      for (const task of phase.tasks) {
        totalTasks++;
        if (task.status === 'passed') {
          completedTasks++;
        }

        for (const subtask of task.subtasks) {
          totalSubtasks++;
          if (subtask.status === 'passed') {
            completedSubtasks++;
          }
        }
      }
    }

    return {
      totalPhases: phases.length,
      completedPhases,
      totalTasks,
      completedTasks,
      totalSubtasks,
      completedSubtasks,
    };
  }

  /**
   * Generates source references for a PRD item from a section.
   * Maps the section back to the requirements document and creates a SourceRef.
   * 
   * @param section - The section this PRD item is derived from
   * @param parsed - The parsed requirements document
   * @param parentSection - Optional parent section for building hierarchical paths
   * @returns Array of source references
   */
  private generateSourceRefs(
    section: ParsedSection,
    parsed: ParsedRequirements,
    parentSection?: ParsedSection
  ): SourceRef[] {
    // Build section path: "Title > Section > Subsection"
    const pathParts: string[] = [];
    if (parsed.title) {
      pathParts.push(parsed.title);
    }
    if (parentSection) {
      pathParts.push(parentSection.title);
    }
    pathParts.push(section.title);
    const sectionPath = pathParts.join(' > ');

    // Calculate hash of section content
    const excerptHash = this.calculateExcerptHash(section.content);

    return [
      {
        sourcePath: parsed.source.path,
        sectionPath,
        excerptHash,
      },
    ];
  }

  /**
   * Calculates SHA-256 hash of excerpt text.
   * 
   * @param excerpt - Text excerpt to hash
   * @returns Hex digest of the hash (64 characters)
   */
  private calculateExcerptHash(excerpt: string): string {
    const hash = createHash('sha256');
    hash.update(excerpt);
    return hash.digest('hex');
  }

  /**
   * Parses JSON response from AI platform into PRD structure.
   * Handles JSON wrapped in markdown code blocks or plain JSON.
   * 
   * @param output - Raw output from AI platform
   * @returns Parsed PRD structure
   */
  private parsePrdJson(output: string): PRD {
    // Try to extract JSON from markdown code blocks
    const jsonMatch = output.match(/```(?:json)?\s*([\s\S]*?)\s*```/);
    const jsonText = jsonMatch ? jsonMatch[1] : output.trim();

    // Try to find JSON object in the text
    const jsonObjectMatch = jsonText.match(/\{[\s\S]*\}/);
    const finalJson = jsonObjectMatch ? jsonObjectMatch[0] : jsonText;

    try {
      const parsed = JSON.parse(finalJson) as PRD;
      
      // Ensure required fields have defaults
      const now = new Date().toISOString();
      if (!parsed.createdAt) parsed.createdAt = now;
      if (!parsed.updatedAt) parsed.updatedAt = now;
      if (!parsed.version) parsed.version = '1.0.0';
      if (!parsed.branchName) parsed.branchName = 'ralph/main';
      if (!parsed.project) parsed.project = this.projectName;
      if (!parsed.description) parsed.description = 'Generated from requirements';
      if (!parsed.metadata) {
        parsed.metadata = this.calculateMetadata(parsed.phases || []);
      }

      // sourceRefs are optional and will be parsed from AI output if present
      // No need to validate or transform them - they should already be in correct format

      return parsed;
    } catch (error) {
      throw new Error(`Failed to parse JSON: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
}

File: puppet-master-rs/src/start_chain/prd_generator.rs
//! PRD generator - creates structured PRD from parsed requirements

use crate::platforms::{PlatformRunner, UsageEvent, UsageTracker};
use crate::types::{
    ExecutionRequest, ItemStatus, PRD, ParsedRequirements, Phase, RequirementsSection, Subtask,
    Task,
};
use anyhow::{Result, anyhow};
use chrono::Utc;
use log::{info, warn};
use std::path::Path;
use std::sync::Arc;
use std::time::{Duration, Instant};

/// Generates PRD from parsed requirements
pub struct PrdGenerator;

fn extract_platform_response_text(raw_output: &str) -> String {
    if let Ok(json) = serde_json::from_str::<serde_json::Value>(raw_output) {
        if let Some(text) = extract_text_from_json(&json) {
            return text;
        }
    }

    if let (Some(start), Some(end)) = (raw_output.find('{'), raw_output.rfind('}')) {
        if start < end {
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(&raw_output[start..=end]) {
                if let Some(text) = extract_text_from_json(&json) {
                    return text;
                }
            }
        }
    }

    raw_output.to_string()
}

fn extract_text_from_json(json: &serde_json::Value) -> Option<String> {
    if let Some(s) = json.as_str() {
        return Some(s.to_string());
    }
    let obj = json.as_object()?;

    for key in ["response", "output", "message", "content", "text"] {
        if let Some(s) = obj.get(key).and_then(|v| v.as_str()) {
            return Some(s.to_string());
        }
    }

    if let Some(result) = obj.get("result") {
        if let Some(s) = result.get("finalResponse").and_then(|v| v.as_str()) {
            return Some(s.to_string());
        }
        if let Some(s) = result.get("response").and_then(|v| v.as_str()) {
            return Some(s.to_string());
        }
        if let Some(s) = result.get("output").and_then(|v| v.as_str()) {
            return Some(s.to_string());
        }
    }

    None
}

fn try_parse_json_value(text: &str) -> Option<serde_json::Value> {
    if let Ok(v) = serde_json::from_str::<serde_json::Value>(text) {
        return Some(v);
    }

    if let (Some(start), Some(end)) = (text.find('{'), text.rfind('}')) {
        if start < end {
            return serde_json::from_str::<serde_json::Value>(&text[start..=end]).ok();
        }
    }

    None
}

impl PrdGenerator {
    /// Generate a PRD from parsed requirements
    pub fn generate(project_name: &str, requirements: &ParsedRequirements) -> Result<PRD> {
        info!("Generating PRD for project: {}", project_name);

        let mut prd = PRD::new(project_name);

        let mut phase_counter = 1;

        // Convert top-level sections to phases
        for section in &requirements.sections {
            let phase = Self::section_to_phase(section, &mut phase_counter)?;
            prd.phases.push(phase);
        }

        Self::inject_default_acceptance_criteria(&mut prd, requirements);
        prd.update_metadata();

        Ok(prd)
    }

    /// Generate a PRD from parsed requirements using an AI platform runner.
    ///
    /// Falls back to rule-based generation if execution or parsing fails.
    pub async fn generate_with_ai(
        project_name: &str,
        requirements: &ParsedRequirements,
        runner: Arc<dyn PlatformRunner>,
        model: &str,
        working_directory: &Path,
        usage_tracker: Option<&UsageTracker>,
    ) -> Result<PRD> {
        // Delegate large documents to multi-pass generator.
        const LARGE_DOC_THRESHOLD: usize = 5000;
        let source_chars: usize = requirements
            .sections
            .iter()
            .map(|s| s.title.len() + s.content.len())
            .sum::<usize>()
            + requirements
                .description
                .as_ref()
                .map(|d| d.len())
                .unwrap_or(0);

        if source_chars >= LARGE_DOC_THRESHOLD {
            let mut multi = super::MultiPassGenerator::new();
            match multi
                .generate_with_ai(
                    requirements,
                    Arc::clone(&runner),
                    model,
                    working_directory,
                    usage_tracker,
                )
                .await
            {
                Ok(mut prd) => {
                    prd.update_metadata();
                    return Ok(prd);
                }
                Err(err) => {
                    warn!(
                        "Multi-pass AI PRD generation failed: {err}; falling back to single-pass"
                    );
                }
            }
        }

        let template = super::PromptTemplates::prd_generation();

        let mut vars = std::collections::HashMap::new();
        vars.insert("project_name".to_string(), project_name.to_string());
        vars.insert(
            "project_description".to_string(),
            Self::extract_description(requirements),
        );
        vars.insert(
            "requirements".to_string(),
            Self::format_requirements_for_prompt(requirements),
        );

        let (system_prompt, user_prompt) = template.render_full(&vars).map_err(|e| anyhow!(e))?;

        let mut prompt = String::new();
        if let Some(system) = system_prompt {
            prompt.push_str(&system);
            prompt.push_str("\n\n");
        }
        prompt.push_str(&user_prompt);
        prompt.push_str(
            "\n\nReturn ONLY a single JSON object matching this Rust PRD schema with camelCase fields: {metadata:{name,description?,version?,createdAt?,updatedAt?},phases:[{id,title,goal?,description?,tasks:[{id,title,description?,subtasks:[{id,taskId,title,description?,acceptanceCriteria:[]}]}]}]}. Do not include markdown fences or any extra text.",
        );

        let request = ExecutionRequest::new(
            runner.platform(),
            model.to_string(),
            prompt,
            working_directory.to_path_buf(),
        )
        .with_timeout(Duration::from_millis(300_000))
        .with_plan_mode(true);

        let start = Instant::now();
        let exec = match runner.execute(&request).await {
            Ok(r) => r,
            Err(err) => {
                warn!("AI PRD generation execution failed: {err}");
                return Self::generate(project_name, requirements);
            }
        };
        let duration_ms = exec
            .duration_ms
            .unwrap_or_else(|| start.elapsed().as_millis() as u64);

        // Track usage (best-effort)
        let mut event = UsageEvent::new(runner.platform())
            .with_model(model.to_string())
            .with_duration(duration_ms)
            .with_success(exec.success);
        if let Some(tokens) = exec.tokens_used {
            event = event.with_tokens(0, tokens);
        }
        if !exec.success {
            event = event.with_error(
                exec.error_message
                    .clone()
                    .unwrap_or_else(|| "AI PRD generation failed".to_string()),
            );
        }
        if let Some(tracker) = usage_tracker {
            let _ = tracker.track(event).await;
        } else if let Ok(tracker) = UsageTracker::default_location() {
            let _ = tracker.track(event).await;
        }

        if !exec.success {
            warn!("AI PRD generation returned unsuccessful result; using heuristic fallback");
            return Self::generate(project_name, requirements);
        }

        let Some(raw_output) = exec.output.as_deref() else {
            warn!("AI PRD generation produced no output; using heuristic fallback");
            return Self::generate(project_name, requirements);
        };

        let response_text = extract_platform_response_text(raw_output);
        let Some(mut json_value) = try_parse_json_value(&response_text) else {
            warn!("Failed to parse AI PRD output as JSON; using heuristic fallback");
            return Self::generate(project_name, requirements);
        };

        // Some models wrap the PRD in a top-level field.
        if let Some(prd_val) = json_value.get("prd").cloned() {
            json_value = prd_val;
        } else if let Some(prd_val) = json_value.get("data").cloned() {
            json_value = prd_val;
        }

        let mut prd: PRD = match serde_json::from_value(json_value) {
            Ok(p) => p,
            Err(err) => {
                warn!("Failed to deserialize AI PRD output: {err}; using heuristic fallback");
                return Self::generate(project_name, requirements);
            }
        };

        if prd.metadata.name.is_empty() {
            prd.metadata.name = project_name.to_string();
        }
        if prd.metadata.created_at.is_none() {
            prd.metadata.created_at = Some(Utc::now());
        }
        prd.metadata.updated_at = Some(Utc::now());

        Self::inject_default_acceptance_criteria(&mut prd, requirements);
        prd.update_metadata();

        Ok(prd)
    }

    fn format_requirements_for_prompt(requirements: &ParsedRequirements) -> String {
        let mut out = String::new();
        for section in &requirements.sections {
            out.push_str(&format!("## {}\n{}\n\n", section.title, section.content));
        }
        if out.trim().is_empty() {
            out = requirements
                .description
                .clone()
                .unwrap_or_else(|| "(no requirements sections provided)".to_string());
        }
        out
    }

    /// Convert a section to a phase
    fn section_to_phase(section: &RequirementsSection, phase_counter: &mut usize) -> Result<Phase> {
        let phase_id = format!("PH-{:03}", *phase_counter);
        *phase_counter += 1;

        let mut phase = Phase {
            id: phase_id.clone(),
            title: section.title.clone(),
            goal: None,
            description: Some(section.content.clone()),
            status: ItemStatus::Pending,
            tasks: Vec::new(),
            iterations: 0,
            evidence: Vec::new(),
            gate_report: None,
            orchestrator_state: None,
            orchestrator_context: None,
            dependencies: Vec::new(),
        };

        // Create a single task from the section content
        let mut task_counter = 1;
        let task = Self::content_to_task(&section.content, &phase_id, &mut task_counter)?;
        phase.tasks.push(task);

        Ok(phase)
    }

    /// Create a task from content
    fn content_to_task(content: &str, phase_id: &str, task_counter: &mut usize) -> Result<Task> {
        let task_id = format!("{}-{:03}", phase_id.replace("PH", "TK"), *task_counter);
        *task_counter += 1;

        let mut subtasks = Vec::new();
        let mut subtask_counter = 1;

        // Split content into lines and create subtasks from bullet items
        for line in content.lines() {
            let trimmed = line.trim();
            if trimmed.starts_with("- ") || trimmed.starts_with("* ") {
                let item_text = trimmed.trim_start_matches("- ").trim_start_matches("* ");
                let subtask = Self::item_to_subtask(item_text, &task_id, &mut subtask_counter)?;
                subtasks.push(subtask);
            }
        }

        // If no subtasks from bullets, create one from the full content
        if subtasks.is_empty() {
            let subtask = Subtask {
                id: format!("{}-{:03}", task_id.replace("TK", "ST"), subtask_counter),
                task_id: task_id.clone(),
                title: "Implementation".to_string(),
                description: Some(content.to_string()),
                criterion: None,
                status: ItemStatus::Pending,
                iterations: 0,
                evidence: Vec::new(),
                plan: None,
                acceptance_criteria: Vec::new(),
                iteration_records: Vec::new(),
            };
            subtasks.push(subtask);
        }

        Ok(Task {
            id: task_id,
            title: "Implementation".to_string(),
            description: Some(content.to_string()),
            status: ItemStatus::Pending,
            subtasks,
            evidence: Vec::new(),
            gate_reports: Vec::new(),
            dependencies: Vec::new(),
            complexity: None,
            task_type: None,
        })
    }

    /// Convert an item to a subtask
    fn item_to_subtask(item: &str, task_id: &str, subtask_counter: &mut usize) -> Result<Subtask> {
        let subtask_id = format!("{}-{:03}", task_id.replace("TK", "ST"), *subtask_counter);
        *subtask_counter += 1;

        let title = if item.len() > 80 {
            item.chars().take(80).collect::<String>() + "..."
        } else {
            item.to_string()
        };

        Ok(Subtask {
            id: subtask_id,
            task_id: task_id.to_string(),
            title,
            description: Some(item.to_string()),
            criterion: None,
            status: ItemStatus::Pending,
            iterations: 0,
            evidence: Vec::new(),
            plan: None,
            acceptance_criteria: Vec::new(),
            iteration_records: Vec::new(),
        })
    }

    /// Inject default acceptance criteria into subtasks that have none.
    ///
    /// Uses the AcceptanceCriteriaInjector to ensure every subtask has machine-verifiable criteria.
    fn inject_default_acceptance_criteria(prd: &mut PRD, _requirements: &ParsedRequirements) {
        let injector = super::AcceptanceCriteriaInjector::default();
        match injector.inject(prd) {
            Ok(result) => {
                info!(
                    "Injected {} criteria into {} subtasks",
                    result.criteria_injected, result.subtasks_without_criteria
                );
            }
            Err(e) => {
                warn!("Failed to inject acceptance criteria: {}", e);
            }
        }
    }

    /// Extract project description from requirements
    fn extract_description(requirements: &ParsedRequirements) -> String {
        requirements
            .description
            .clone()
            .or_else(|| requirements.sections.first().map(|s| s.content.clone()))
            .unwrap_or_else(|| "Project generated from requirements".to_string())
    }

    /// Save PRD to file
    pub async fn save_prd(prd: &PRD, path: &std::path::Path) -> Result<()> {
        let json = serde_json::to_string_pretty(prd)?;

        // Ensure directory exists
        if let Some(parent) = path.parent() {
            tokio::fs::create_dir_all(parent).await?;
        }

        tokio::fs::write(path, json).await?;
        info!("PRD saved to {:?}", path);

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_prd() {
        let mut requirements = ParsedRequirements::new("Test Project");
        requirements.sections.push(RequirementsSection::new(
            "Phase 1",
            "Phase description\n- Item 1\n- Item 2",
        ));

        let prd = PrdGenerator::generate("Test Project", &requirements).unwrap();

        assert_eq!(prd.metadata.name, "Test Project");
        assert_eq!(prd.phases.len(), 1);
        assert_eq!(prd.phases[0].title, "Phase 1");
        
        // Verify acceptance criteria were injected
        for phase in &prd.phases {
            for task in &phase.tasks {
                for subtask in &task.subtasks {
                    assert!(!subtask.acceptance_criteria.is_empty(),
                        "Subtask {} should have acceptance_criteria injected", subtask.id);
                }
            }
        }
    }

    #[test]
    fn test_acceptance_criteria_injection() {
        let mut requirements = ParsedRequirements::new("Test Project");
        requirements.sections.push(RequirementsSection::new(
            "Implementation",
            "Build the feature\n- Write code\n- Add tests",
        ));

        let prd = PrdGenerator::generate("Test Project", &requirements).unwrap();

        // Verify subtasks have prefixed acceptance criteria
        let subtask = &prd.phases[0].tasks[0].subtasks[0];
        assert!(!subtask.acceptance_criteria.is_empty());
        
        // Check for prefixed format
        let has_prefixed = subtask.acceptance_criteria.iter().any(|c| {
            c.starts_with("command:") || c.starts_with("file_exists:") || c.starts_with("regex:")
        });
        assert!(has_prefixed, "Generated PRD should have prefixed acceptance criteria");
    }

    #[test]
    fn test_section_to_phase() {
        let section = RequirementsSection::new("Test Phase", "Description\n- Item 1");

        let mut counter = 1;
        let phase = PrdGenerator::section_to_phase(&section, &mut counter).unwrap();

        assert_eq!(phase.id, "PH-001");
        assert_eq!(phase.title, "Test Phase");
        assert!(!phase.tasks.is_empty());
    }
}
